streams/common/src/main/java/com/hortonworks/streamline/streams/common/event/correlation/CorrelatedEventsGrouper.java

```diff
@@ -35,27 +37,29 @@ public CorrelatedEventsGrouper(List<EventInformation> events) {
     }
 
     /**
-     * Returns the group of correlated events per source
+     * Returns the group of correlated events per source where the group of related source events
+     * produces the same downstream events
      *
      * @param sources the set of source component names
      * @return the group of correlated events per source
      */
-    public Set<Set<String>> groupByRelatedSourceEvents(Set<String> sources) {
-        Set<Set<String>> result = new LinkedHashSet<>();
+    public List<Set<String>> groupByRelatedSourceEvents(Set<String> sources) {
+        Multimap<Set<String>, Set<String>> allEventsToSourceEvents = HashMultimap.create();
         Stream<String> rootEventIds = events.stream().filter(e -> e != null && e.getRootIds().isEmpty())
                 .map(EventInformation::getEventId);
         rootEventIds.forEach(rootEventId -> {
             Set<String> relatedSourceEvents = new LinkedHashSet<>();
```

Comment:
We don't need to collect related source events from allRelatedEvents, given that we are checking all the root events. So all the root events which have same related events should be grouped together, and naturally they should be correlated even we don't check it, because A, B, C are grouped, their same related events should have A, B, C.

Reply:
So it will work if you group root events which has same related events. If the group is [A, B, C], regardless of selection (one of A, B, C) the same events DAG will be shown.

Reply:
I made some more changes to group all root events that result in same "allEvents". Please check.